src/modules/billing/services/supplementary-billing-service/supplementary-processor.js

```diff
@@ -0,0 +1,302 @@
+  };
+  return hashers.createMd5Hash(JSON.stringify(data));
+};
+
+const xor = (a, b) => a ? !b : b;
+
+/**
+ * Maps an array of transactions to a data structure which includes a summed total
+ *
+ * The summed total sums:
+ * - billable days in the group (for annual)
+ * - volumes in the group (for TPT)
+ *
+ * Depending on whether the transaction is a credit, and whether it is in the current
+ * batch, it is negated.  This means where a transaction in a sent batch has the
+ * same billable days/volume as one in the current batch, they will sum to zero
+ * and the transaction in the current batch will then be deleted.
+ *
+ * If the sum is non-zero, then transactions will remain in the current batch.
+ *
+ * @param {String} batchId
+ * @param {Array<Object>} transactions
+ * @return {Object} data structure containing { transactions: [], sum }
+ */
+const mapTransactionGroup = (batchId, transactions) => transactions.reduce((acc, transaction) => {
+  const isCurrentBatch = transaction.billingBatchId === batchId;
+  const propertyKey = transaction.isTwoPartTariffSupplementary ? 'volume' : 'billableDays';
+  let value = new Decimal(transaction[propertyKey]);
+  if (xor(transaction.isCredit, !isCurrentBatch)) {
+    value = value.negated();
+  }
+  return {
+    transactions: [...acc.transactions, {
+      ...transaction,
+      isCurrentBatch,
+      action: null
+    }],
+    sum: acc.sum.plus(value)
+  };
+}, {
+  transactions: [],
+  sum: new Decimal(0)
+});
+
+/**
+ * Predicate to check if this transaction should be considered.
+ *
+ * We only wish to consider standard and compensation charges.
+ * Minumum charge correction transactions are omitted.
+ *
+ * @param {Object} transaction
+ * @return {Boolean}
+ */
+const isValidTransaction = transaction =>
+  ['standard', 'compensation'].includes(transaction.chargeType);
+
+/**
+ * Predicate to check if transaction is part of the current batch
+ *
+ * @param {Object} transaction
+ * @return {Boolean}
+ */
+const isCurrentBatchTransaction = transaction => transaction.isCurrentBatch;
+
+/**
+ * Predicate to check if transaction is part of an existing batch
+ *
+ * @param {Object} transaction
+ * @return {Boolean}
+ */
+const isExistingBatchTransaction = negate(isCurrentBatchTransaction);
+
+/**
+ * Deletes all transactions in the given group that are
+ * in the current batch
+ *
+ * @param {Object} group
+ * @param {Array<Object>} group.transactions
+ * @return {Promise}
+ */
+const markCurrentBatchTransactionsForDeletion = group => {
+  group.transactions
+    .filter(isCurrentBatchTransaction)
+    .forEach(transaction => {
+      transaction.action = actions.deleteTransaction;
+    });
+};
+
+/**
+ * Gets a key for pairing transactions within a group which have the same:
+ * - Charge period
+ * - Billable days (for annual transactions)
+ * - Volume (for TPT transactions)
+ *
+ * @param {Object} transaction
+ * @return {String}
+ */
+const getPairGroupingKey = transaction => {
```

Comment:
when the transactions are grouped and filtered - do they not need to be grouped by charge element Id or invoice licence Id?

Reply:
No, we are looking for:
* Transactions that have come from a mostly identical charge element, but not with the same ID
* The historical transactions won't be on the same invoice or invoice licence as the current batch ones as they are spread across multiple billing batches